vshard/replicaset.lua

```diff
@@ -75,6 +75,141 @@ local fiber_cond_wait = util.fiber_cond_wait
 local future_wait = util.future_wait
 local gsc = util.generate_self_checker
 
+--
+-- vconnect is asynchronous vshard greeting, saved inside netbox connection.
+-- It stores future object and additional info, needed for its work.
+-- Future is initialized, when the connection is established (inside
+-- netbox_on_connect). The connection cannot be considered "connected"
+-- until vconnect is properly validated.
+--
+local function conn_vconnect_set(conn)
+    assert(conn.replica ~= nil)
+    local is_named = conn.replica.id == conn.replica.name
+    if not is_named then
+        -- Nothing to do. Check is not needed.
+        return
+    end
+    -- Update existing vconnect. Fiber condition cannot be dropped,
+    -- somebody may already waiting on it.
+    if conn.vconnect then
+        -- Connections are preserved during reconfiguration,
+        -- identification may be changed.
+        conn.vconnect.is_named = is_named
+        if not conn.vconnect.future then
+            return
+        end
+        -- Future object must be updated. Old result is irrelevant.
+        conn.vconnect.future:discard()
+        conn.vconnect.future = nil
+        return
+    end
+    -- Create new vconnect.
+    conn.vconnect = {
+        -- Whether the connection is done to the named replica.
+        is_named = is_named,
+        -- Used to wait for the appearance of vconnect.future object,
+        -- if call is done before the connection is established.
+        future_cond = fiber.cond(),
+    }
+end
+
+--
+-- Initialize future object. Should be done, when connection is established
+-- (inside netbox_on_connect).
+--
+local function conn_vconnect_start(conn)
+    local vconn = conn.vconnect
+    if not vconn then
+        -- Nothing to do. Check is not needed.
+        return
+    end
+
+    local opts = {is_async = true}
+    vconn.future = conn:call('vshard.storage._call', {'info'}, opts)
+    vconn.future_cond:broadcast()
+end
+
+--
+-- Check, that future is ready, and its result is expected.
+-- The function doesn't yield.
+-- @retval true; The correct response is received.
+-- @retval nil, ...; Response is not received or validation error happened.
+--
+local function conn_vconnect_check(conn)
+    local vconn = conn.vconnect
+    -- conn.vconnect may be nil, if connection was created on old version
+    -- and the storage was reloaded to a new one. It's also nil, when
+    -- all checks were already done.
+    if not vconn then
+        return true
+    end
+    -- Nothing to do, but wait in such case.
+    if not vconn.future or not vconn.future:is_ready() then
+        return nil, lerror.vshard(lerror.code.VHANDSHAKE_NOT_COMPLETE,
+                                  conn.replica.id)
+    end
```

Comment:
You need to check the error type first. Otherwise this condition will be true also when `error.code` equals to something from `box.error.`

Reply:
I don't get it. When some error happens, `vconn.future:is_ready()` equals to true and we'll filter it out below. When there's no result, it will be in this check:

```
tarantool> function test() box.error('error') end
tarantool> f = c:eval('test', {}, {is_async = true})
tarantool> f:is_ready()
---
- true
...

tarantool> f:result()
---
- null
- Execute access to universe '' is denied for user 'guest'
```

Reply:
Sorry, I left the comment in a wrong place. I've meant this:
```Lua
local function conn_vconnect_check_or_close(conn)
    local ok, err = conn_vconnect_check(conn)
    if not ok and err.code ~= lerror.code.VHANDSHAKE_NOT_COMPLETE then
        conn:close()
    end
    return ok, err
end
```
If `err` is a `ClientError` with code `ER_WAL_IO` (= 40, like `VHANDSHAKE_NOT_COMPLETE`), then your condition will pass. So you need to check the error type to be equal to `ShardingError` before you can check the code.

Reply:
Now I see and agree. `err.type` is also checked now.